### PR TITLE
feat(icon): add support for icon purpose

### PR DESCRIPTION
**Pull request by hans00**
Let icon maskable.
https://developer.mozilla.org/en-US/docs/Web/Manifest/icons

### CODE REVIEW

File: docs/modules/icon.md [hunk 1]
```diff
@@ -48,3 +48,12 @@ Will return an empty string when no icon in the given size is available (eg. whe
 - Default: '$icon'
 
 Name of property for accessible icons.
+
+**purpose**
+- Default: null
+
+Array of icon purpose.
```
Comment (pi0) [inline docs/modules/icon.md]: Can you please improve docs by putting more description or links?

File: lib/icon/module.js [hunk 2]
```diff
@@ -113,7 +114,8 @@ async function generateIcons (options) {
   options._assetIcons = options.sizes.map(size => ({
     src: joinUrl(options.publicPath, options._icons[size]),
     sizes: `${size}x${size}`,
-    type: 'image/png'
+    type: 'image/png',
+    purpose: options.purpose ? options.purpose.join(' ') : undefined
```
Comment (ricardogobbosouza) [inline lib/icon/module.js]: - Add validation, the types allowed: `'badge', 'maskable', 'any'`
- Also recommend checking if `purpose` is string, put in array

File: docs/modules/icon.md (hunk 1 above)
Comment (hans00) [reply docs/modules/icon.md]: I added some description, please review.

File: lib/icon/module.js [hunk 3]
```diff
@@ -45,6 +46,18 @@ async function run (pwa, _emitAssets) {
...
+    if (typeof options.purpose === 'string') {
+      options.purpose = [options.purpose]
+    } else if (options.purpose.constructor === Array) {
+      const validPurpose = ['badge', 'maskable', 'any']
+      options.purpose = options.purpose.filter(item => validPurpose.includes(item))
+    } else {
+      options.purpose = null
+    }
+  }
+
```
Comment (ricardogobbosouza) [inline lib/icon/module.js]: If we pass `purpose: 'foo'` will be accepted, try this:

```js
if (options.purpose) {
  const purpose = Array.isArray(options.purpose) ? options.purpose : [options.purpose]
  const validPurpose = ['badge', 'maskable', 'any']
  options.purpose = purpose.filter(item => validPurpose.includes(item))
}
```

**Comment by ulissesferreira** [comment]
Hey guys, just wanted to say maskable icons are officially [out](https://www.youtube.com/watch?v=L0OB0_bO5I0) what do we need to get this rolling? Can I help?

File: lib/icon/module.js [hunk 4]
```diff
@@ -45,6 +46,13 @@ async function run (pwa, _emitAssets) {
     return
   }
 
+  // Verify purpose
+  if (options.purpose) {
+    const purpose = Array.isArray(options.purpose) ? options.purpose : [options.purpose]
+    const validPurpose = ['badge', 'maskable', 'any']
+    options.purpose = purpose.filter(item => validPurpose.includes(item))
```
Comment (pi0) [inline lib/icon/module.js]: Silently filtering invalid items is not a good idea. Maybe we can show a warning instead? (one idea can be keeping length before filter and warn if length changed)

File: lib/icon/module.js [hunk 5]
```diff
@@ -109,11 +117,15 @@ async function generateIcons (options) {
     options._icons[size] = `${options.targetDir}/icon_${size}.${options.iconHash}.png`
   }
 
+  // Generate _purpose
+  options._purpose = options.purpose ? options.purpose.join(' ') : undefined
```
Comment (pi0) [inline lib/icon/module.js]: Do we need `_purpose` anywhere else? If not, can we just use `const purpose =  ...`?

File: lib/icon/module.js (hunk 4 above)
Comment (ricardogobbosouza) [reply lib/icon/module.js]: warn user is a good idea

**Comment by TomSmith27** [comment]
Can someone merge this change in, as without this icons have a white background on some phones and don't look very good

File: lib/icon/module.js (hunk 4 above)
Comment (pi0) [reply lib/icon/module.js]: 0ff11d41434daa3cb37459606c82b1a5fbdeb293

File: lib/icon/module.js (hunk 5 above)
Comment (pi0) [reply lib/icon/module.js]: 5638999a09ed341bc1a6533cd84246b06998519d

**Comment by pi0** [comment]
I've fixed remaining tasks. Thanks @hans00 and @ricardogobbosouza :heart:

**Pull request merged**